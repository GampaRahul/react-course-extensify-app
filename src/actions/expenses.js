import {v4 as uuid} from 'uuid';
// ADD_EXPENCE

export const addExpense =  ({
    description = '', 
    note = '', 
    amount = 0, 
    createdAt = 0
    }={}
    ) => ({
    type: 'ADD_EXPENCE',
    expence:{
                id: uuid(),
                description,
                note,
                amount,
                createdAt
    }
});
// REMOVE_EXPENCE
export const removeExpense = ({id}={}) =>({
    type:'REMOVE_EXPENCE',
    id
});
// EDIT_EXPENCE
export const editExpense = (id,update) =>({
    type:'EDIT_EXPENCE',
    id,
    update
});