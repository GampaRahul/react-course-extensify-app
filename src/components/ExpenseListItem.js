import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({expense, dispatch}=props) => (
    <div>
        <h3><Link to={`/edit/${expense.id}`}>{expense.description}</Link></h3>
        <p>
        â‚¹{expense.amount} 
        - 
        {moment(expense.createdAt).format("MMMM Do, YYYY")}</p>
    </div>
);

export default ExpenseListItem;
