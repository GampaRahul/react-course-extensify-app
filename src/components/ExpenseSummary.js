import React from 'react';
import {connect} from 'react-redux';
import expensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export const ExpenseSummary = ({expenses}) => {
    const expenseText = expenses.length===1?'expense':'expenses';
    return(
        <div>
            <h1>viewing {expenses.length} {expenseText} totalling â‚¹{expensesTotal(expenses)}</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses,state.filters)
    };
}

export default connect(mapStateToProps)(ExpenseSummary);