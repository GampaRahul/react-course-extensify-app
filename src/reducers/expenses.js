//expenseReducer
const expenceReduserDefaultState = [];

const expenceReduser = (state = expenceReduserDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter((expence) => expence.id !== action.id); 
        case 'EDIT_EXPENSE':
            return state.map((expence) =>{
                if(expence.id === action.id){
                    return {
                        ...expence,
                        ...action.update
                    }
                }
                else return expence
            });
        default:
            return state;
    }

};

export default expenceReduser;