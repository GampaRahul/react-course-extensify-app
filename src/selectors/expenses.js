import moment from 'moment';
//getVisibleExpences
//if start date or end date is undefined we gonna exclude them in filter
const getVisibleExpences = (expenses, {text, sortBy, startDate, endDate}) =>{
    return expenses.filter((expense) =>{
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate? endDate.isSameOrAfter(createdAtMoment,'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expense1, expense2) =>{
        if(sortBy == 'date'){
            return expense1.createdAt < expense2.createdAt ?1:-1
        }
        else if(sortBy == 'amount'){
            return expense1.amount < expense2.amount ?1:-1
        }
    });
};


export default getVisibleExpences;