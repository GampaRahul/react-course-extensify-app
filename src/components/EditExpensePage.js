import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpenses} from '../actions/expenses';
import {startRemoveExpense} from '../actions/expenses';



export class EditExpensePage extends React.Component{
  onSubmit = (expense) =>{
    this.props.startEditExpenses(this.props.expense.id,expense);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.startRemoveExpense({id:this.props.expense.id});
    this.props.history.push('/');
  };

  render(){

    return(
      <div>
        <ExpenseForm 
        expense = {this.props.expense}
        onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) =>{
  return{
    expense : state.expenses.find((expense) => 
      expense.id===props.match.params.id
    )
  };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    startEditExpenses: (id, expense) => dispatch(startEditExpenses(id,expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
