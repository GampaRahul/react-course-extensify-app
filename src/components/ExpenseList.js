import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {//<h1>Expense List</h1>
        }
        {
            props.expenses.length === 0?(
                <p>No Expenses</p>
            ):(
                props.expenses.map((expense) => <ExpenseListItem expense={expense} key={expense.id}/>)
            )
        }
    </div>

);


//this specifies what all we need from store, so connect() can map that to props
const mapStateToProps = (state) => {
    return {
        // selectExpenses return the expenses in a sorted order, depending on the filter
        expenses: selectExpenses(state.expenses, state.filters)
    };
};


// this is HOC which we use in other components
export default connect(mapStateToProps)(ExpenseList);
