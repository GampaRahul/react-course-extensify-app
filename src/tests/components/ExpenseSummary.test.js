import {shallow} from 'enzyme';
import React from 'react';
import {ExpenseSummary} from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';


test('should render summary of 2 expenses', () => {
    const exp=[expenses[0],expenses[1]];
    const wrapper = shallow(<ExpenseSummary expenses={exp}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render summary of 1 expense', () => {
    const exp=[expenses[0]];
    const wrapper = shallow(<ExpenseSummary expenses={exp}/>);
    expect(wrapper).toMatchSnapshot();
});