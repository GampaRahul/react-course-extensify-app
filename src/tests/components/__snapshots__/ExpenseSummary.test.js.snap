// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render summary of 1 expense 1`] = `
<div>
  <h1>
    viewing 
    1
     
    expense
     totalling ₹
    10
  </h1>
</div>
`;

exports[`should render summary of 2 expenses 1`] = `
<div>
  <h1>
    viewing 
    2
     
    expenses
     totalling ₹
    10010
  </h1>
</div>
`;
